src/Microsoft.IdentityModel.Tokens/MaxCapacityRefreshCache.cs

```diff
@@ -0,0 +1,137 @@
+        /// </summary>
+        /// <param name="sizeLimit">The max number of items the cache is capable of holding.</param>
+        /// <param name="comparer">The string comparer.</param>
+        internal MaxCapacityRefreshCache(
+            long sizeLimit,
+            IEqualityComparer<TKey> comparer = null)
+        {
+            RecycleLock = new object();
+            SizeLimit = sizeLimit;
+            Cache = new ConcurrentDictionary<TKey, TValue>(comparer ?? EqualityComparer<TKey>.Default);
+        }
+
+        internal bool Contains(TKey key)
+        {
+            return Cache.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// Gets the value corresponding to <paramref name="value"/> from the cache.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The cached value.</param>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
+        internal bool TryGetValue(TKey cacheKey, out TValue value)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryGetValue(cacheKey, out value);           
+        }
+
+        /// <summary>
+        /// Sets the cache value.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The value.</param>
+        /// <returns><paramref name="value"/> if it was successfully set, null otherwise.</returns>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
+        /// <remarks>This value will not be set if the cache already contains a key with the same name as <paramref name="cacheKey"/>.</remarks>
+        internal TValue SetValue(TKey cacheKey, TValue value)
+        {
+            RecycleCacheIfNeeded();
+            if (Cache.TryAdd(cacheKey, value))
+                return value;
+            else
+                return null;
+        }
+
+        /// <summary>
+        /// Removes a particular key from the cache.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        internal bool TryRemove(TKey cacheKey)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryRemove(cacheKey, out _);
+        }
+
+        /// <summary>
+        /// Removes the value corresponding to <paramref name="cacheKey"/>.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="removedValue">The removed value.</param>
+        internal bool TryRemove(TKey cacheKey, out TValue removedValue)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryRemove(cacheKey, out removedValue);
+        }
+
+        /// <summary>
+        /// Recycles the cache if needed.
+        /// </summary>
+        private void RecycleCacheIfNeeded()
+        {
+            if (SizeLimit == Cache.Keys.Count)
+            {
+                lock (RecycleLock)
```

Comment:
I'm not sure if we really need to lock here.

Reply:
ConcurrentDictionary.Clear should do a "table level" lock. it does in 4.8, need to verify that it does in 4.5

Reply:
How did you know it does a table level lock in 4.8? I can't seem to find this information in the official documentation.

Reply:
I **think** the idea is it is implied. That is the only way Clear can be done in a thread safe way. But I agree, it isn't in the docs and leaves you to hope.... or look at the reference source: https://github.com/microsoft/referencesource/blob/NetFramework48ZDP/mscorlib/system/collections/Concurrent/ConcurrentDictionary.cs#L615

Reply:
Thank you for confirming that a table level lock occurs in 4.5 as well! I've gone ahead and removed the lock from this caching implementation. :)